src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -978,36 +1157,41 @@ public void RemoveFile(string fullPath)
                         throw new ArgumentException($"'{fullPath}' is not a source file of this project.");
                     }
 
-                    _documentPathsToDocumentIds.Remove(fullPath);
-
                     _project._documentFileChangeContext.StopWatchingFile(_project._fileWatchingTokens[documentId]);
                     _project._fileWatchingTokens.Remove(documentId);
 
-                    // There are two cases:
-                    // 
-                    // 1. This file is actually been pushed to the workspace, and we need to remove it (either
-                    //    as a part of the active batch or immediately)
-                    // 2. It hasn't been pushed yet, but is contained in _documentsAddedInBatch
-                    if (_documentAlreadyInWorkspace(_project._workspace.CurrentSolution, documentId))
+                    RemoveFileInternal(documentId, fullPath);
+                }
+            }
+
+            private void RemoveFileInternal(DocumentId documentId, string fullPath)
```

Comment:
consier renaming to RemoveFileWhileHoldingLock.  Also, assert the lock has been entered.